### ISSUE TITLE
Add comments to the code to help newcomers

### ISSUE THREAD

**Issue by thefzsalam** [sdroege/gobject-subclass#6]
Please add descriptions to the structs, traits and macros in gobject-subclass code.
Thanks :)

**Comment by sdroege** [comment]
That's definitely a good idea! Until that point, what are you trying to do? Maybe I can link you to some example code then for the time being :)

**Comment by thefzsalam** [comment]
I need to implement GListModel and gtk_list_box_bind_model method for gtk-rs.
I'm not very experienced in GObject, but would love to learn more. :)

**Comment by thefzsalam** [comment]
Once I understand how things work, I would like to write comments for gobject-subclass too.

**Comment by sdroege** [comment]
Ah that one is going to be tricky in any case because of all the `gpointer` / `void *` usage in the C API.

In gtk-rs, this probably has to become a generic type `ListModel<T: StaticType + FromGlibPtrNone + ToGlibPtr>`, as completely manual bindings. Same thing goes for `GListStore`. It's very bindings unfriendly API, I don't think any of this can be autogenerated.

As a first step you'll have to start with the bindings part in the `gio` and `gtk` bindings. And can then make the `GListStore` work first. After that implementing everything for subclassing should be relatively straightforward, but the other part has to happen first.

**Comment by thefzsalam** [comment]
I will try to implement `bind_model` accepting `IsA<ListStore>` instead of `ListModel`. Then if that works out, I will create wrappers for the interface `gio::GListModel`.
But how to wrap `GtkListBoxCreateWidgetFunc`?

**Comment by sdroege** [comment]
You would have to write manual bindings for that too, around a `FnMut`. There are some like that in `glib/src/source.rs` IIRC, take those as an example.

But the important part here is that you somehow need to make the `gpointer` everywhere into a generic type parameter. You can take a look at `gstreamer/src/iterator.rs` for some code doing something similar.

**Comment by thefzsalam** [comment]
What is the use of `struct ClassInitToken(());` and `struct TypeInitToken(());` ?

**Comment by sdroege** [comment]
To make it impossible to call the functions that need them as parameter from arbitrary places

**Comment by itsfarseen** [comment]
Could you give some examples from where it is impossible to call functions with `ClassInitToken` as a parameter?

**Comment by sdroege** [comment]
`ClassInitToken` can only be created from inside this crate, and you only get a reference to one passed to the class init function. So you can only (with safe code) call functions taken a `ClassInitToken` from inside the class init function.

**Comment by itsfarseen** [comment]
But it is declared public here: https://github.com/gtk-rs/gobject-subclass/blob/f42869195b1022223885889c99f375f9f8e4e4f4/gobject-subclass/src/object.rs#L67
And `object.rs` is imported as `pub mod object` here: https://github.com/gtk-rs/gobject-subclass/blob/f42869195b1022223885889c99f375f9f8e4e4f4/gobject-subclass/src/lib.rs#L25

**Comment by sdroege** [comment]
Yes but the constructor is private, it would otherwise have been `pub struct ClassInitToken(pub ());`

**Comment by itsfarseen** [comment]
What does `Ext` stand for in the trait names like `ObjectClassExt` or `ListBoxExt`?

**Comment by sdroege** [comment]
They're extension traits for the type in their name, and its subclasses

**Comment by itsfarseen** [comment]
https://github.com/gtk-rs/examples/raw/master/src/bin/listbox_model.rs is a very well documented example implementation. 
Please add a link to this in Readme.md

**Comment by sdroege** [comment]
@happycoder97 Do you want to send a PR to add this to the `README.md`?

**Comment by itsfarseen** [comment]
Done :)

**Comment by sdroege** [comment]
The code is now merged into https://github.com/gtk-rs/glib/tree/master/src/subclass in a quit different and cleaned up form, and that has documentation/comments on everything now.

Please open a new issue if there's still something missing there. This repo here is obsolete :)
